src/bundles/peer-locations.js

```diff
@@ -300,44 +140,67 @@ class PeerLocationResolver {
       ...opts.cache
     })
 
+    this.failedAddrs = HLRU(500)
+
+    this.queue = new PQueue({
+      concurrency: opts.concurrency,
+      autoStart: true
+    })
+
     this.geoipLookupPromises = {}
   }
 
-  async lookupWithCaches (peerId, addr, getIpfs) {
-    const ipv4Tuple = Multiaddr(addr).stringTuples().find(isNonHomeIPv4)
-    if (!ipv4Tuple) {
-      throw new Error(`Unable to resolve location for non-IPv4 address ${addr}`)
-    }
+  async findLocations (peers, getIpfs) {
+    const res = {}
 
-    var ipv4Addr = ipv4Tuple[1]
+    for (const p of peers) {
+      const peerId = p.peer.toB58String()
+      const addr = p.addr.toString()
 
-    // maybe we have it cached by ipv4 address already, check that.
-    let location = await this.geoipCache.get(ipv4Addr)
-    if (location) {
-      return location
-    }
+      const ipv4Tuple = Multiaddr(addr).stringTuples().find(isNonHomeIPv4)
+      if (!ipv4Tuple) {
+        continue
+      }
+
+      const ipv4Addr = ipv4Tuple[1]
+      if (this.failedAddrs.has(ipv4Addr)) {
+        continue
+      }
+
+      // maybe we have it cached by ipv4 address already, check that.
+      const location = await this.geoipCache.get(ipv4Addr)
+      if (location) {
+        res[peerId] = location
+        continue
+      }
 
-    // no ip address cached. are we looking it up already?
-    var locPromise = this.geoipLookupPromises[ipv4Addr]
+      // no ip address cached. are we looking it up already?
+      var locPromise = this.geoipLookupPromises[ipv4Addr]
+      if (locPromise) {
+        continue
+      }
+
+      this.queue.add(async () => {
+        return new Promise(resolve => {
+          const ipfs = getIpfs()
 
-    if (!locPromise) {
-      locPromise = new Promise((resolve, reject) => {
-        const ipfs = getIpfs()
-        geoip.lookup(ipfs, ipv4Addr, (err, data) => {
-          // save it in the caches, and remove the promise.
-          this.geoipCache.set(ipv4Addr, data)
-          delete this.geoipLookupPromises[ipv4Addr]
+          geoip.lookup(ipfs, ipv4Addr, (err, data) => {
+            delete this.geoipLookupPromises[ipv4Addr]
 
-          if (err) return reject(err)
-          resolve(data)
+            if (err) {
+              // mark this one as failed so we don't retry again
+              this.failedAddrs.set(ipv4Addr, 1)
```

Comment:
`true` looks better than a magic number :)
```suggestion
              this.failedAddrs.set(ipv4Addr, true)
```

Reply:
Thanks!